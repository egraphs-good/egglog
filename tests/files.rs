use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
