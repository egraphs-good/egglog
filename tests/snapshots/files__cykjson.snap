---
source: tests/files.rs
expression: serialized
---
nodes:
  Unit-0:
    op: ()
    children: []
    eclass: Unit-0
    cost: 0
  i64-5871781006564002453:
    op: "1"
    children: []
    eclass: i64-5871781006564002453
    cost: 0
  i64-17147757364762811680:
    op: "160"
    children: []
    eclass: i64-17147757364762811680
    cost: 0
  String-8150770311509600722:
    op: "\"H9\""
    children: []
    eclass: String-8150770311509600722
    cost: 0
  P-15144575588694754941:
    op: P
    children:
      - i64-5871781006564002453
      - i64-17147757364762811680
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-13250620236881053665:
    op: "285"
    children: []
    eclass: i64-13250620236881053665
    cost: 0
  P-6660603375870406635:
    op: P
    children:
      - i64-5871781006564002453
      - i64-13250620236881053665
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-16108815799063573221:
    op: "785"
    children: []
    eclass: i64-16108815799063573221
    cost: 0
  P-1273666883823903385:
    op: P
    children:
      - i64-5871781006564002453
      - i64-16108815799063573221
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-10081113907175992806:
    op: "718"
    children: []
    eclass: i64-10081113907175992806
    cost: 0
  P-17922521199970013583:
    op: P
    children:
      - i64-5871781006564002453
      - i64-10081113907175992806
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-4053412015288412391:
    op: "651"
    children: []
    eclass: i64-4053412015288412391
    cost: 0
  P-7652032519244375412:
    op: P
    children:
      - i64-5871781006564002453
      - i64-4053412015288412391
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-7482786487199436520:
    op: "328"
    children: []
    eclass: i64-7482786487199436520
    cost: 0
  P-2005925516484039893:
    op: P
    children:
      - i64-5871781006564002453
      - i64-7482786487199436520
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-1455084595311856105:
    op: "261"
    children: []
    eclass: i64-1455084595311856105
    cost: 0
  P-14537644614317802246:
    op: P
    children:
      - i64-5871781006564002453
      - i64-1455084595311856105
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-13406718123246169834:
    op: "706"
    children: []
    eclass: i64-13406718123246169834
    cost: 0
  P-14614914964862122408:
    op: P
    children:
      - i64-5871781006564002453
      - i64-13406718123246169834
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-17667493649174738220:
    op: "380"
    children: []
    eclass: i64-17667493649174738220
    cost: 0
  P-7073739528995426935:
    op: P
    children:
      - i64-5871781006564002453
      - i64-17667493649174738220
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-2442583535694516531:
    op: "679"
    children: []
    eclass: i64-2442583535694516531
    cost: 0
  P-15556507817656699869:
    op: P
    children:
      - i64-5871781006564002453
      - i64-2442583535694516531
      - String-8150770311509600722
    eclass: Unit-0
    cost: 1
  i64-17823591535539854389:
    op: "801"
    children: []
    eclass: i64-17823591535539854389
    cost: 0
  String-3098702898132563177:
    op: "\"VAL\""
    children: []
    eclass: String-3098702898132563177
    cost: 0
  B-16791568458799218522:
    op: B
    children:
      - i64-17823591535539854389
      - i64-5871781006564002453
      - String-3098702898132563177
    eclass: "0"
    cost: 100000
  B-15144575588694754941:
    op: B
    children:
      - i64-5871781006564002453
      - i64-17147757364762811680
      - String-8150770311509600722
    eclass: "1"
    cost: 100000
  B-6660603375870406635:
    op: B
    children:
      - i64-5871781006564002453
      - i64-13250620236881053665
      - String-8150770311509600722
    eclass: "1"
    cost: 100000
  B-1273666883823903385:
    op: B
    children:
      - i64-5871781006564002453
      - i64-16108815799063573221
      - String-8150770311509600722
    eclass: "1"
    cost: 100000
  B-17922521199970013583:
    op: B
    children:
      - i64-5871781006564002453
      - i64-10081113907175992806
      - String-8150770311509600722
    eclass: "1"
    cost: 100000
  B-7652032519244375412:
    op: B
    children:
      - i64-5871781006564002453
      - i64-4053412015288412391
      - String-8150770311509600722
    eclass: "1"
    cost: 100000
  B-2005925516484039893:
    op: B
    children:
      - i64-5871781006564002453
      - i64-7482786487199436520
      - String-8150770311509600722
    eclass: "1"
    cost: 100000
  B-14537644614317802246:
    op: B
    children:
      - i64-5871781006564002453
      - i64-1455084595311856105
      - String-8150770311509600722
    eclass: "1"
    cost: 100000
  B-14614914964862122408:
    op: B
    children:
      - i64-5871781006564002453
      - i64-13406718123246169834
      - String-8150770311509600722
    eclass: "1"
    cost: 100000
  B-7073739528995426935:
    op: B
    children:
      - i64-5871781006564002453
      - i64-17667493649174738220
      - String-8150770311509600722
    eclass: "1"
    cost: 100000
  String-3400952766172528380:
    op: "\"H7\""
    children: []
    eclass: String-3400952766172528380
    cost: 0
  NT-2176608565598431445:
    op: NT
    children:
      - String-3400952766172528380
      - T-15139977626647034397
      - T-2994503827742450628
    eclass: "12"
    cost: 1
  String-7017119389452091262:
    op: "\"H8\""
    children: []
    eclass: String-7017119389452091262
    cost: 0
  NT-17905997656708914419:
    op: NT
    children:
      - String-7017119389452091262
      - T-1293589485465454534
      - B-15144575588694754941
    eclass: "13"
    cost: 1
  NT-9361484887233955811:
    op: NT
    children:
      - String-7017119389452091262
      - T-6732138773774815835
      - B-6660603375870406635
    eclass: "14"
    cost: 1
  String-1845757586690957117:
    op: "\"ELM\""
    children: []
    eclass: String-1845757586690957117
    cost: 0
  NT-11442316754243898537:
    op: NT
    children:
      - String-1845757586690957117
      - NT-17905997656708914419
      - T-17023076937514119005
    eclass: "15"
    cost: 1
  NT-12200805522705888519:
    op: NT
    children:
      - String-1845757586690957117
      - NT-9361484887233955811
      - T-17023076937514119005
    eclass: "16"
    cost: 1
  String-2688846101539080723:
    op: "\"MEM\""
    children: []
    eclass: String-2688846101539080723
    cost: 0
  NT-3071556225764271360:
    op: NT
    children:
      - String-2688846101539080723
      - NT-2176608565598431445
      - T-6732138773774815835
    eclass: "17"
    cost: 1
  NT-12320915076839643495:
    op: NT
    children:
      - String-2688846101539080723
      - NT-2176608565598431445
      - T-1293589485465454534
    eclass: "18"
    cost: 1
  String-14420720653836506280:
    op: "\"PR\""
    children: []
    eclass: String-14420720653836506280
    cost: 0
  NT-7239000051061034411:
    op: NT
    children:
      - String-14420720653836506280
      - NT-2176608565598431445
      - T-6732138773774815835
    eclass: "19"
    cost: 1
  NT-16436385273695213892:
    op: NT
    children:
      - String-14420720653836506280
      - NT-2176608565598431445
      - T-1293589485465454534
    eclass: "20"
    cost: 1
  String-2278989304945598269:
    op: "\"H2\""
    children: []
    eclass: String-2278989304945598269
    cost: 0
  NT-1174069112454411311:
    op: NT
    children:
      - String-2278989304945598269
      - T-3828771413392206835
      - NT-12320915076839643495
    eclass: "21"
    cost: 1
  String-5764174078010485202:
    op: "\"str\""
    children: []
    eclass: String-5764174078010485202
    cost: 0
  End-6732138773774815835:
    op: End
    children:
      - String-3098702898132563177
      - String-5764174078010485202
    eclass: Unit-0
    cost: 1
  String-14022551318073603175:
    op: "\"num\""
    children: []
    eclass: String-14022551318073603175
    cost: 0
  End-1293589485465454534:
    op: End
    children:
      - String-3098702898132563177
      - String-14022551318073603175
    eclass: Unit-0
    cost: 1
  String-4653898077614134440:
    op: "\"true\""
    children: []
    eclass: String-4653898077614134440
    cost: 0
  End-5621862773378465073:
    op: End
    children:
      - String-3098702898132563177
      - String-4653898077614134440
    eclass: Unit-0
    cost: 1
  String-11635955084574487655:
    op: "\"false\""
    children: []
    eclass: String-11635955084574487655
    cost: 0
  End-860357767210813382:
    op: End
    children:
      - String-3098702898132563177
      - String-11635955084574487655
    eclass: Unit-0
    cost: 1
  String-10525679084178136893:
    op: "\"null\""
    children: []
    eclass: String-10525679084178136893
    cost: 0
  End-18196825840524014236:
    op: End
    children:
      - String-3098702898132563177
      - String-10525679084178136893
    eclass: Unit-0
    cost: 1
  End-17023076937514119005:
    op: End
    children:
      - String-1845757586690957117
      - String-5764174078010485202
    eclass: Unit-0
    cost: 1
  End-13537892164449232072:
    op: End
    children:
      - String-1845757586690957117
      - String-14022551318073603175
    eclass: Unit-0
    cost: 1
  End-2506436815954674823:
    op: End
    children:
      - String-1845757586690957117
      - String-4653898077614134440
    eclass: Unit-0
    cost: 1
  End-11151295930950116552:
    op: End
    children:
      - String-1845757586690957117
      - String-11635955084574487655
    eclass: Unit-0
    cost: 1
  End-15081399883100223986:
    op: End
    children:
      - String-1845757586690957117
      - String-10525679084178136893
    eclass: Unit-0
    cost: 1
  String-12045811881167970109:
    op: "\"{\""
    children: []
    eclass: String-12045811881167970109
    cost: 0
  getString-5871781006564002453:
    op: getString
    children:
      - i64-5871781006564002453
    eclass: String-12045811881167970109
    cost: 1
  i64-11743562013128004906:
    op: "2"
    children: []
    eclass: i64-11743562013128004906
    cost: 0
  getString-11743562013128004906:
    op: getString
    children:
      - i64-11743562013128004906
    eclass: String-5764174078010485202
    cost: 1
  String-9682590569330013287:
    op: "\":\""
    children: []
    eclass: String-9682590569330013287
    cost: 0
  i64-17615343019692007359:
    op: "3"
    children: []
    eclass: i64-17615343019692007359
    cost: 0
  getString-17615343019692007359:
    op: getString
    children:
      - i64-17615343019692007359
    eclass: String-9682590569330013287
    cost: 1
  String-8970483904696565630:
    op: "\"[\""
    children: []
    eclass: String-8970483904696565630
    cost: 0
  i64-5040379952546458196:
    op: "4"
    children: []
    eclass: i64-5040379952546458196
    cost: 0
  getString-5040379952546458196:
    op: getString
    children:
      - i64-5040379952546458196
    eclass: String-8970483904696565630
    cost: 1
  i64-10912160959110460649:
    op: "5"
    children: []
    eclass: i64-10912160959110460649
    cost: 0
  getString-10912160959110460649:
    op: getString
    children:
      - i64-10912160959110460649
    eclass: String-12045811881167970109
    cost: 1
  i64-16783941965674463102:
    op: "6"
    children: []
    eclass: i64-16783941965674463102
    cost: 0
  getString-16783941965674463102:
    op: getString
    children:
      - i64-16783941965674463102
    eclass: String-5764174078010485202
    cost: 1
  i64-4208978898528913939:
    op: "7"
    children: []
    eclass: i64-4208978898528913939
    cost: 0
  getString-4208978898528913939:
    op: getString
    children:
      - i64-4208978898528913939
    eclass: String-9682590569330013287
    cost: 1
  i64-10080759905092916392:
    op: "8"
    children: []
    eclass: i64-10080759905092916392
    cost: 0
  getString-10080759905092916392:
    op: getString
    children:
      - i64-10080759905092916392
    eclass: String-5764174078010485202
    cost: 1
  String-13999176396412444477:
    op: "\",\""
    children: []
    eclass: String-13999176396412444477
    cost: 0
  i64-15952540911656918845:
    op: "9"
    children: []
    eclass: i64-15952540911656918845
    cost: 0
  getString-15952540911656918845:
    op: getString
    children:
      - i64-15952540911656918845
    eclass: String-13999176396412444477
    cost: 1
  i64-3377577844511369682:
    op: "10"
    children: []
    eclass: i64-3377577844511369682
    cost: 0
  getString-3377577844511369682:
    op: getString
    children:
      - i64-3377577844511369682
    eclass: String-5764174078010485202
    cost: 1
  test1-0:
    op: test1
    children: []
    eclass: "0"
    cost: 9223372036854776000
  T-17144903266680156499:
    op: T
    children:
      - String-8150770311509600722
      - String-13999176396412444477
    eclass: "1"
    cost: 1
  String-8548939647272503827:
    op: "\"H0\""
    children: []
    eclass: String-8548939647272503827
    cost: 0
  T-3828771413392206835:
    op: T
    children:
      - String-8548939647272503827
      - String-12045811881167970109
    eclass: "2"
    cost: 1
  T-1293589485465454534:
    op: T
    children:
      - String-3098702898132563177
      - String-14022551318073603175
    eclass: "3"
    cost: 1
  T-13537892164449232072:
    op: T
    children:
      - String-1845757586690957117
      - String-14022551318073603175
    eclass: "4"
    cost: 1
  String-18339137145156034365:
    op: "\"H11\""
    children: []
    eclass: String-18339137145156034365
    cost: 0
  T-2994503827742450628:
    op: T
    children:
      - String-18339137145156034365
      - String-9682590569330013287
    eclass: "5"
    cost: 1
  String-15975915833318077543:
    op: "\"H3\""
    children: []
    eclass: String-15975915833318077543
    cost: 0
  T-5066758862541156673:
    op: T
    children:
      - String-15975915833318077543
      - String-8970483904696565630
    eclass: "6"
    cost: 1
  String-2136319993900342971:
    op: "\"H10\""
    children: []
    eclass: String-2136319993900342971
    cost: 0
  T-15139977626647034397:
    op: T
    children:
      - String-2136319993900342971
      - String-5764174078010485202
    eclass: "7"
    cost: 1
  T-6732138773774815835:
    op: T
    children:
      - String-3098702898132563177
      - String-5764174078010485202
    eclass: "8"
    cost: 1
  T-17023076937514119005:
    op: T
    children:
      - String-1845757586690957117
      - String-5764174078010485202
    eclass: "9"
    cost: 1
  String-16385772629911559997:
    op: "\"H4\""
    children: []
    eclass: String-16385772629911559997
    cost: 0
  String-3810809562766010834:
    op: "\"]\""
    children: []
    eclass: String-3810809562766010834
    cost: 0
  T-11860242468977441006:
    op: T
    children:
      - String-16385772629911559997
      - String-3810809562766010834
    eclass: "10"
    cost: 1
  String-14853952372091147432:
    op: "\"H1\""
    children: []
    eclass: String-14853952372091147432
    cost: 0
  Prod-12429448979527149223:
    op: Prod
    children:
      - String-3098702898132563177
      - String-8548939647272503827
      - String-14853952372091147432
    eclass: Unit-0
    cost: 1
  Prod-7477076631530229353:
    op: Prod
    children:
      - String-3098702898132563177
      - String-2278989304945598269
      - String-14853952372091147432
    eclass: Unit-0
    cost: 1
  Prod-6055220555749997347:
    op: Prod
    children:
      - String-3098702898132563177
      - String-15975915833318077543
      - String-16385772629911559997
    eclass: Unit-0
    cost: 1
  String-6174030874603967656:
    op: "\"H5\""
    children: []
    eclass: String-6174030874603967656
    cost: 0
  Prod-4866665062179759367:
    op: Prod
    children:
      - String-3098702898132563177
      - String-6174030874603967656
      - String-16385772629911559997
    eclass: Unit-0
    cost: 1
  String-10923848419941039998:
    op: "\"H6\""
    children: []
    eclass: String-10923848419941039998
    cost: 0
  Prod-14904433591509596411:
    op: Prod
    children:
      - String-2688846101539080723
      - String-10923848419941039998
      - String-2688846101539080723
    eclass: Unit-0
    cost: 1
  Prod-1204630798424365285:
    op: Prod
    children:
      - String-2688846101539080723
      - String-3400952766172528380
      - String-3098702898132563177
    eclass: Unit-0
    cost: 1
  Prod-1384773258010952990:
    op: Prod
    children:
      - String-14420720653836506280
      - String-3400952766172528380
      - String-3098702898132563177
    eclass: Unit-0
    cost: 1
  Prod-12324591201326854254:
    op: Prod
    children:
      - String-1845757586690957117
      - String-7017119389452091262
      - String-1845757586690957117
    eclass: Unit-0
    cost: 1
  Prod-9470472476404999529:
    op: Prod
    children:
      - String-1845757586690957117
      - String-8548939647272503827
      - String-14853952372091147432
    eclass: Unit-0
    cost: 1
  Prod-5732534202318183167:
    op: Prod
    children:
      - String-1845757586690957117
      - String-2278989304945598269
      - String-14853952372091147432
    eclass: Unit-0
    cost: 1
root_eclasses: []
class_data:
  Unit-0:
    type: Unit
  i64-5871781006564002453:
    type: i64
  i64-17147757364762811680:
    type: i64
  String-8150770311509600722:
    type: String
  i64-13250620236881053665:
    type: i64
  i64-16108815799063573221:
    type: i64
  i64-10081113907175992806:
    type: i64
  i64-4053412015288412391:
    type: i64
  i64-7482786487199436520:
    type: i64
  i64-1455084595311856105:
    type: i64
  i64-13406718123246169834:
    type: i64
  i64-17667493649174738220:
    type: i64
  i64-2442583535694516531:
    type: i64
  "0":
    type: tree
  i64-17823591535539854389:
    type: i64
  String-3098702898132563177:
    type: String
  "1":
    type: tree
  "12":
    type: tree
  String-3400952766172528380:
    type: String
  "7":
    type: tree
  "5":
    type: tree
  "13":
    type: tree
  String-7017119389452091262:
    type: String
  "3":
    type: tree
  "14":
    type: tree
  "8":
    type: tree
  "15":
    type: tree
  String-1845757586690957117:
    type: String
  "9":
    type: tree
  "16":
    type: tree
  "17":
    type: tree
  String-2688846101539080723:
    type: String
  "18":
    type: tree
  "19":
    type: tree
  String-14420720653836506280:
    type: String
  "20":
    type: tree
  "21":
    type: tree
  String-2278989304945598269:
    type: String
  "2":
    type: tree
  String-5764174078010485202:
    type: String
  String-14022551318073603175:
    type: String
  String-4653898077614134440:
    type: String
  String-11635955084574487655:
    type: String
  String-10525679084178136893:
    type: String
  String-12045811881167970109:
    type: String
  i64-11743562013128004906:
    type: i64
  String-9682590569330013287:
    type: String
  i64-17615343019692007359:
    type: i64
  String-8970483904696565630:
    type: String
  i64-5040379952546458196:
    type: i64
  i64-10912160959110460649:
    type: i64
  i64-16783941965674463102:
    type: i64
  i64-4208978898528913939:
    type: i64
  i64-10080759905092916392:
    type: i64
  String-13999176396412444477:
    type: String
  i64-15952540911656918845:
    type: i64
  i64-3377577844511369682:
    type: i64
  String-8548939647272503827:
    type: String
  "4":
    type: tree
  String-18339137145156034365:
    type: String
  "6":
    type: tree
  String-15975915833318077543:
    type: String
  String-2136319993900342971:
    type: String
  "10":
    type: tree
  String-16385772629911559997:
    type: String
  String-3810809562766010834:
    type: String
  String-14853952372091147432:
    type: String
  String-6174030874603967656:
    type: String
  String-10923848419941039998:
    type: String

